File: bin/controller.go
package bin

import (
	"encoding/json"
	"net/http"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"syscall"

	"golang.org/x/sys/unix"
)

func GetNameClient() {
	path := S99xkeen
	if _, err := os.Stat(S24xray); err == nil { path = S24xray }
	InitFile = path
	b, _ := os.ReadFile(path)
	name := "xray"
	if strings.Contains(string(b), `name_client="mihomo"`) { name = "mihomo" }
	ClientMutex.Lock()
	CurrentCore = clientTypes[name]
	ClientMutex.Unlock()
}

func getPid(name string) int {
	ents, _ := os.ReadDir("/proc")
	for _, e := range ents {
		if !e.IsDir() || e.Name()[0] < '0' || e.Name()[0] > '9' { continue }
		if b, _ := os.ReadFile("/proc/" + e.Name() + "/comm"); strings.TrimSpace(string(b)) == name {
			if p, err := strconv.Atoi(e.Name()); err == nil { return p }
		}
	}
	return 0
}

func getCoreVersion(name string) string {
	arg := "version"
	if name == "mihomo" { arg = "-v" }
	out, err := exec.Command(name, arg).Output()
	if err != nil { return "" }
	s := string(out)
	parts := strings.Fields(s)
	if name == "xray" && len(parts) > 1 { return "v" + parts[1] }
	if name == "mihomo" && len(parts) > 2 { return parts[2] }
	return ""
}

func ControlHandler(w http.ResponseWriter, r *http.Request) {
	DebugLog("ControlHandler: method=%s", r.Method)
	if r.Method == http.MethodGet {
		ClientMutex.RLock()
		cur := CurrentCore.Name
		ClientMutex.RUnlock()

		if getPid(cur) == 0 {
			alt := "mihomo"
			if cur == "mihomo" { alt = "xray" }
			if getPid(alt) > 0 {
				ClientMutex.Lock()
				CurrentCore = clientTypes[alt]
				ClientMutex.Unlock()
				cur = alt
			}
		}

		cores, running := []string{}, false
		versions := make(map[string]string)
		for _, c := range []string{"xray", "mihomo"} {
			if _, err := os.Stat("/opt/sbin/" + c); err == nil {
				cores = append(cores, c)
				versions[c] = getCoreVersion(c)
			}
			if getPid(c) > 0 { running = true }
		}

		jsonResponse(w, map[string]any{
			"success":     true,
			"cores":       cores,
			"currentCore": cur,
			"running":     running,
			"versions":    versions,
		}, 200)
		return
	}

	var req struct{ Action, Core string }
	if json.NewDecoder(r.Body).Decode(&req) != nil {
		DebugLog("ControlHandler: JSON decode error")
		jsonResponse(w, Response{Success: false, Error: "JSON error"}, 400)
		return
	}
	DebugLog("ControlHandler POST: action=%s core=%s", req.Action, req.Core)

	f, _ := os.OpenFile(ErrorLog, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	defer f.Close()

	ClientMutex.RLock()
	curName := CurrentCore.Name
	ClientMutex.RUnlock()

	if curName == "mihomo" && (req.Action == "start" || req.Action == "hardRestart") { os.Truncate(ErrorLog, 0) }

	switch req.Action {
	case "switchCore":
		DebugLog("ControlHandler: switching core to %s", req.Core)
		ClientMutex.RLock()
		oldCore := CurrentCore.Name
		if oldCore == req.Core {
			ClientMutex.RUnlock()
			DebugLog("ControlHandler: already on %s", req.Core)
			jsonResponse(w, Response{Success: true, Data: "Already " + req.Core}, 200)
			return
		}
		ClientMutex.RUnlock()

		exec.Command(InitFile, "stop").Run()

		if _, err := os.Stat(S99xkeen); err == nil { InitFile = S99xkeen } else { InitFile = S24xray }
		if data, err := os.ReadFile(InitFile); err == nil {
			out := strings.Replace(string(data), `name_client="`+oldCore+`"`, `name_client="`+req.Core+`"`, 1)
			os.WriteFile(InitFile, []byte(out), 0755)
		}

		ClientMutex.Lock()
		if c, ok := clientTypes[req.Core]; ok { CurrentCore = c }
		ClientMutex.Unlock()

		if CurrentCore.Name != "xray" { os.Truncate(ErrorLog, 0) }

		cmd := exec.Command(InitFile, "start")
		cmd.Stdout, cmd.Stderr = f, f
		cmd.Run()
		DebugLog("ControlHandler: core switched to %s", req.Core)

	case "softRestart":
		DebugLog("ControlHandler: soft restarting %s", req.Core)
		if pid := getPid(req.Core); pid > 0 {
			DebugLog("ControlHandler: killing pid %d", pid)
			syscall.Kill(pid, syscall.SIGKILL)
		}

		cmd := exec.Command(req.Core)
		cmd.Env = append(os.Environ(), coreEnvs[req.Core]...)
		cmd.SysProcAttr = &syscall.SysProcAttr{Credential: &syscall.Credential{Gid: 11111}}
		cmd.Stdout, cmd.Stderr = f, f

		if cmd.Start() != nil {
			DebugLog("ControlHandler: start failed")
			jsonResponse(w, Response{Success: false, Error: "Start fail"}, 500)
			return
		}

		limit := uint64(10000)
		if runtime.GOARCH == "arm64" { limit = 40000 }
		unix.Prlimit(cmd.Process.Pid, unix.RLIMIT_NOFILE, &unix.Rlimit{Cur: limit, Max: limit}, nil)
		go cmd.Wait()
		DebugLog("ControlHandler: soft restart done, pid=%d", cmd.Process.Pid)

	case "start", "stop", "hardRestart":
		arg := req.Action
		if req.Action == "hardRestart" { arg = "restart" }
		DebugLog("ControlHandler: executing %s %s", InitFile, arg)
		c := exec.Command(InitFile, arg)
		c.Stdout, c.Stderr = f, f
		c.Run()
		DebugLog("ControlHandler: %s %s done", InitFile, arg)

	default:
		DebugLog("ControlHandler: bad action %s", req.Action)
		jsonResponse(w, Response{Success: false, Error: "Bad action"}, 400)
		return
	}
	jsonResponse(w, Response{Success: true}, 200)
}

File: src/controller.rs
use axum::{extract::State, response::{IntoResponse, Json}};
use serde::Deserialize;
use std::{collections::HashMap, path::Path, os::unix::fs::PermissionsExt, fs::Permissions, io::Read};
use nix::{sys::signal::{kill, Signal}, unistd::Pid};
use tokio::{fs::{self, set_permissions}, process::Command};
use crate::types::*;

#[derive(Deserialize)]
pub struct ControlReq {
    action: String,
    #[serde(default)]
    core: String
}

fn get_core_info(name: &str) -> CoreInfo {
    match name {
        "mihomo" => CoreInfo { name: "mihomo".into(), conf_dir: MIHOMO_CONF.into(), is_json: false },
        _ => CoreInfo { name: "xray".into(), conf_dir: XRAY_CONF.into(), is_json: true }
    }
}

pub fn get_pid(name: &str) -> Option<i32> {
    let mut buf = Vec::new();
    std::fs::read_dir("/proc").ok()?.filter_map(|e| {
        let p = e.ok()?.path();
        let pid = p.file_name()?.to_str()?.parse::<i32>().ok()?;
        buf.clear();
        std::fs::File::open(p.join("comm")).ok()?.read_to_end(&mut buf).ok()?;
        (buf.strip_suffix(b"\n")? == name.as_bytes()).then_some(pid)
    }).next()
}

pub async fn soft_restart(core: &str) {
    if let Some(pid) = get_pid(core) { let _ = kill(Pid::from_raw(pid), Signal::SIGKILL); }
    let mut cmd = Command::new(core);
    match core {
        "xray" => cmd.envs([("XRAY_LOCATION_CONFDIR", XRAY_CONF), ("XRAY_LOCATION_ASSET", XRAY_ASSET)]),
        _ => cmd.env("CLASH_HOME_DIR", MIHOMO_CONF)
    };
    unsafe { cmd.pre_exec(|| { nix::libc::setsid(); nix::libc::setgid(11111); Ok(()) }); }

    if let Ok(f) = std::fs::OpenOptions::new().create(true).append(true).open(ERROR_LOG) {
        cmd.stdout(f.try_clone().unwrap()).stderr(f);
    }
    if let Ok(mut child) = cmd.spawn() {
        if let Some(id) = child.id() {
            let limit = if cfg!(target_arch = "aarch64") { 40000 } else { 10000 };
            unsafe { nix::libc::prlimit(id as i32, nix::libc::RLIMIT_NOFILE, &nix::libc::rlimit { rlim_cur: limit, rlim_max: limit }, std::ptr::null_mut()); }
        }
        tokio::spawn(async move { let _ = child.wait().await; });
    }
}

pub async fn get_control(State(state): State<AppState>) -> impl IntoResponse {
    let mut core = state.core.read().unwrap().clone();
    if get_pid(&core.name).is_none() {
        let alt = if core.name == "mihomo" { "xray" } else { "mihomo" };
        core = if get_pid(alt).is_some() {
            get_core_info(alt)
        } else {
            let path = state.init_file.read().unwrap();
            let conf = std::fs::read_to_string(&*path).unwrap_or_default();
            get_core_info(if conf.contains("name_client=\"mihomo\"") { "mihomo" } else { "xray" })
        };
        *state.core.write().unwrap() = core.clone();
    }

    let mut versions = HashMap::new();
    let mut cores = Vec::new();
    let mut running = false;

    for c in ["xray", "mihomo"] {
        if Path::new(&format!("/opt/sbin/{}", c)).exists() {
            cores.push(c.to_string());
            let out = Command::new(c)
                .arg(if c == "mihomo" { "-v" } else { "version" })
                .output().await.ok();

            if let Some(o) = out {
                let s = String::from_utf8_lossy(&o.stdout);
                let parts: Vec<&str> = s.split_whitespace().collect();
                let v = match c {
                    "xray" if parts.len() > 1 => {
                         if parts[1].starts_with('v') { parts[1].into() } else { format!("v{}", parts[1]) }
                    },
                    "mihomo" if parts.len() > 2 => parts[2].into(),
                    _ => "?".into()
                };
                versions.insert(c.to_string(), v);
            }
        }
        if get_pid(c).is_some() { running = true; }
    }

    Json(serde_json::json!({ "success": true, "cores": cores, "currentCore": core.name, "running": running, "versions": versions }))
}

pub async fn post_control(State(state): State<AppState>, Json(req): Json<ControlReq>) -> impl IntoResponse {
    match req.action.as_str() {
        "switchCore" => {
            let old = state.core.read().unwrap().name.clone();
            if old == req.core { return Json(ApiResponse { success: true, error: None, data: None }); }

            let init_file = state.init_file.read().unwrap().clone();
            let _ = Command::new(&init_file).arg("stop").status().await;

            let new_init_file = if Path::new(S99XKEEN).exists() { S99XKEEN.to_string() } else { S24XRAY.to_string() };
            *state.init_file.write().unwrap() = new_init_file.clone();

            if let Ok(content) = fs::read_to_string(&new_init_file).await {
                let new_content = content.replace(&format!("name_client=\"{}\"", old), &format!("name_client=\"{}\"", req.core));
                let _ = fs::write(&new_init_file, new_content).await;
                let permissions = Permissions::from_mode(0o755);
                let _ = set_permissions(&new_init_file, permissions).await;
            }

            *state.core.write().unwrap() = get_core_info(&req.core);

            if req.core != "xray" { let _ = fs::write(ERROR_LOG, b"").await; }

            if let Ok(f) = std::fs::OpenOptions::new().create(true).append(true).open(ERROR_LOG) {
                let _ = Command::new(&new_init_file).arg("start").stdout(f.try_clone().unwrap()).stderr(f).status().await;
            } else {
                let _ = Command::new(&new_init_file).arg("start").status().await;
            }
        },
        "softRestart" => soft_restart(&req.core).await,
        a if ["start", "stop", "hardRestart"].contains(&a) => {
            let arg = match a {
                "start" => "start",
                "stop" => "stop",
                _ => "restart",
            };

            let cur_name = state.core.read().unwrap().name.clone();
            if cur_name == "mihomo" && (a == "start" || a == "hardRestart") {
                let _ = fs::write(ERROR_LOG, b"").await;
            }

            let init_file = state.init_file.read().unwrap().clone();
            if let Ok(f) = std::fs::OpenOptions::new().create(true).append(true).open(ERROR_LOG) {
                let s = Command::new(&init_file).arg(arg).stdout(f.try_clone().unwrap()).stderr(f).status().await;
                if matches!(s, Ok(c) if !c.success()) {
                    return Json(ApiResponse { success: false, error: Some("Process error".into()), data: None });
                }
            } else {
                let _ = Command::new(&init_file).arg(arg).status().await;
            }
        },
        _ => return Json(ApiResponse { success: false, error: Some("Bad action".into()), data: None }),
    }
    Json(ApiResponse::<()> { success: true, error: None, data: None })
}